discovery-provider/src/queries/search_es.py

```diff
@@ -0,0 +1,412 @@
+                    {"index": ES_PLAYLISTS},
+                    album_dsl(search_str, current_user_id, True),
+                ]
+            )
+
+    for dsl in mdsl:
+        if "query" in dsl:
+            dsl["size"] = limit
+            dsl["from"] = offset
+
+    mfound = esclient.msearch(searches=mdsl)
+
+    response: Dict = {
+        "tracks": [],
+        "saved_tracks": [],
+        "users": [],
+        "followed_users": [],
+        "playlists": [],
+        "saved_playlists": [],
+        "albums": [],
+        "saved_albums": [],
+    }
+
+    if do_tracks:
+        response["tracks"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["saved_tracks"] = pluck_hits(mfound["responses"].pop(0))
+
+    if do_users:
+        response["users"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["followed_users"] = pluck_hits(mfound["responses"].pop(0))
+
+    if do_playlists:
+        response["playlists"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["saved_playlists"] = pluck_hits(mfound["responses"].pop(0))
+
+    if do_albums:
+        response["albums"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["saved_albums"] = pluck_hits(mfound["responses"].pop(0))
+
+    # hydrate users, saves, reposts
+    item_keys = []
+    user_ids = set()
+    if current_user_id:
+        user_ids.add(current_user_id)
+
+    # collect keys for fetching
+    for k in [
+        "tracks",
+        "saved_tracks",
+        "playlists",
+        "saved_playlists",
+        "albums",
+        "saved_albums",
+    ]:
+        for item in response[k]:
+            item_keys.append(item_key(item))
+            user_ids.add(item.get("owner_id", item.get("playlist_owner_id")))
+
+    # fetch users
+    users_by_id = {}
+    current_user = None
+
+    if user_ids:
+        users_mget = esclient.mget(index=ES_USERS, ids=list(user_ids))
+        users_by_id = {d["_id"]: d["_source"] for d in users_mget["docs"] if d["found"]}
+        if current_user_id:
+            current_user = users_by_id.get(str(current_user_id))
+        for id, user in users_by_id.items():
+            users_by_id[id] = populate_user_metadata_es(user, current_user)
+
+    # fetch followed saves + reposts
+    # TODO: instead of limit param (20) should do an agg to get 3 saves / reposts per item_key
+    (follow_saves, follow_reposts) = fetch_followed_saves_and_reposts(
+        current_user_id, item_keys, 20
+    )
+
+    # tracks: finalize
+    for k in ["tracks", "saved_tracks"]:
+        tracks = response[k]
+        hydrate_user(tracks, users_by_id)
+        hydrate_saves_reposts(tracks, follow_saves, follow_reposts)
+        response[k] = transform_tracks(tracks, users_by_id, current_user)
+
+    # users: finalize
+    for k in ["users", "followed_users"]:
+        users = drop_copycats(response[k])
```

Comment:
just so i understand, we're well-protected here b/c of overfetch yeah?

Reply:
No that's a good catch.  I will add some over-fetching to user `limit`